### PR TITLE
Allow comments & whitespace in gitignore files & fix issue where global ignore rules may not be written.

### PR DESCRIPTION
**Pull request by bencorlett**
As outlined in my commits, I typically don't install Magento to `htdocs` relative to this project, I will install it directly to the same directory as Composer.

I usually have a well structured gitignore file that works well for me. It has comments and various things in it.

I refactored the `GitIgnore` class to allow for whitespace, comments etc without breaking the formatting.

I also fixed an issue whereby if you had directories that were ignored entirely that differed from the state of the gitignore file, it won't have been updated. Only a change in file entries could have previously triggered a rewrite of the gitignore file, not just ignore rules.

### CODE REVIEW

**Comment by AydinHassan** [comment]
Hey, the third issue was intentional, was that causing you an issue?

File: src/GitIgnore.php [hunk 1]
```diff
@@ -113,7 +113,20 @@ public function wipeOut()
     public function __destruct()
     {
         if ($this->hasChanges) {
-            file_put_contents($this->gitIgnoreLocation, implode("\n", array_flip($this->lines)));
+            file_put_contents($this->gitIgnoreLocation, implode("\n", $this->lines));
+        }
+    }
+
+    /**
+     * Add entires to for all directories ignored entirely.
```
Comment (AydinHassan) [inline src/GitIgnore.php]: typo in entries

File: test/GitIgnoreTest.php [hunk 2]
```diff
@@ -91,13 +119,22 @@ public function testAddEntryDoesNotAddFileOrDirectoryIfItIsInsideAnIgnoredDirect
 
     public function testIgnoreDirectoriesAreNotWrittenIfNoEntriesAreAdded()
```
Comment (AydinHassan) [inline test/GitIgnoreTest.php]: Yeah so this test shows that the feature was intentional. My thoughts were that why would you add these folders if you are not actually adding anything else? I'm not against the change, but if we implement then we need to remove this test but it's not doing as it actually describes now.

Comment (bencorlett) [inline test/GitIgnoreTest.php]: My reasoning is, say, you decide to change the ignored directories in your `composer.json` file... The only way to have that actually take effect is to empty those lines from your `.gitignore` file, remove Magento's files and reinstall.

This hasn't actually hit me though so I'm not really worried if we don't include it. I just saw it as an opportunity to solve an issue before somebody gets stumped by that edge-case.

Comment (bencorlett) [inline test/GitIgnoreTest.php]: With that test too, it appears that what you have there is still valid. In that test, there were no entries at the start. You provided no further entries, so the file was not written.

Correct me if I'm wrong, but what was happening before was that no entries were written **even if** you added new ignored directories. That didn't make this test incorrect, but it didn't capture the edge case that existed. Even with the new changes, this test is still valid.

**Comment by AydinHassan** [comment]
OK I'm going to go ahead and merge this, thanks!
